tests/lib/prepare.sh

```diff
@@ -192,6 +192,54 @@ update_core_snap_for_classic_reexec() {
     done
 }
 
+prepare_memory_limit_override() {
+    # set up memory limits for snapd bu default unless explicit requested not to
+    # or the system is known to be problematic
+    local set_limit=1
+
+    case "$SPREAD_SYSTEM" in
+        ubuntu-core-16-*|ubuntu-core-18-*|ubuntu-16.04-*)
+            # the tests on UC16 and UC18 have demonstrated that the memory limit
+            # state claimed by systemd may be out of sync with actual memory
+            # controller setting for the snapd.service cgroup
+            set_limit=0
+            ;;
+        amazon-linux-*)
+            # similar issues have been observed on Amazon Linux 2
+            set_limit=0
+            ;;
+        *)
+            if [ -n "${SNAPD_NO_MEMORY_LIMIT:-}" ]; then
+                set_limit=0
+            fi
+            ;;
+    esac
+
+    if [ "$set_limit" = "0" ]; then
+        # make sure the file does not exist then
+        rm -f /etc/systemd/system/snapd.service.d/memory-max.conf
+    else
+        mkdir -p /etc/systemd/system/snapd.service.d
+        # Use MemoryMax to set the memory limit for snapd.service, that is the
+        # main snapd process and its subprocesses executing within the same
+        # cgroup. If snapd hits the memory limit, it will get killed by
+        # oom-killer which will be caught in restore_project_each in
+        # prepare-restore.sh.
+        #
+        # This ought to set MemoryMax, but on systems with older systemd we need to
+        # use MemoryLimit, which is deprecated and replaced by MemoryMax now, but
+        # systemd is backwards compatible so the limit is still set.
+        cat <<EOF > /etc/systemd/system/snapd.service.d/memory-max.conf
+[Service]
+MemoryLimit=100M
```

Comment:
why not set both like we do for quota groups?

Reply:
Missed this one. I'm waiting for spread run to complete. If it lands soonish I'll open a followup.